### PR TITLE
AMDGPU: Legalize atomicrmw fadd for v2f16/v2bf16 for local memory

### PR DESCRIPTION
**Pull request by arsenm**
Make this legal for gfx940 and gfx12

### CODE REVIEW

**Comment by arsenm** [comment]
* **#95396** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95396?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#95395** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95395?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#95394** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95394?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#95393** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/95393?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @arsenm and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Target/AMDGPU/SIISelLowering.cpp [hunk 1]
```diff
@@ -15931,6 +15931,16 @@ static OptimizationRemark emitAtomicRMWLegalRemark(const AtomicRMWInst *RMW) {
          << " operation at memory scope " << MemScope;
 }
 
+static bool isHalf2OrBFloat2(Type *Ty) {
+  if (FixedVectorType *VT = dyn_cast<FixedVectorType>(Ty)) {
```
Comment (shiltian) [inline llvm/lib/Target/AMDGPU/SIISelLowering.cpp:15935]: nit. I think LLVM standard suggests to use `auto` where cast is used.

```suggestion
  if (auto *VT = dyn_cast<FixedVectorType>(Ty)) {
```

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp [hunk 2]
```diff
@@ -301,6 +301,9 @@ static const LLT V10S16 = LLT::fixed_vector(10, 16);
 static const LLT V12S16 = LLT::fixed_vector(12, 16);
 static const LLT V16S16 = LLT::fixed_vector(16, 16);
 
+static const LLT V2F16 = LLT::fixed_vector(2, LLT::float16());
+static const LLT V2BF16 = V2F16; // FIXME
```
Comment (shiltian) [inline llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:305]: what happens here?

**Review by shiltian — APPROVED** [review]
LGTM with nits

File: llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/AMDGPULegalizerInfo.cpp:305]: bfloat currently has no globalisel representation

**Comment by arsenm** [comment]
### Merge activity

* **Jun 15, 3:51 AM EDT**: @arsenm started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/95393).

**Pull request merged**